### ISSUE TITLE
Auto ARCH

### ISSUE THREAD

**Issue by azev77** [s-broda/ARCHModels.jl#60]
Hi and thanks for the awesome package! 
It's becoming common to automate time-series models. 
[Forecast](https://github.com/robjhyndman/forecast).R has [auto.arima](https://www.rdocumentation.org/packages/forecast/versions/8.12/topics/auto.arima)() which returns the "best" Arima model according to some criteria defined by the user. 

It would be awesome if your package has some options to select the "best" ARCH model. 
[Here](https://www.r-bloggers.com/automatic-armagarch-selection-in-parallel/) is some discussion about doing this in [R](https://gist.github.com/ivannp/5198580). 

Thanks again for your awesome package.

**Comment by s-broda** [comment]
Thanks for the praise!

Are you thinking about automatic lag order selection? The package can already to that, see the documentation [here](https://s-broda.github.io/ARCHModels.jl/stable/usage/#Model-selection-1). The short of it is that the function `selectmodel` can automatically select the best lag order for both the ARCH specification and the ARMA mean specification. By default it uses the BIC criterion, but you can pass you own. The function is multihreaded, so it will be very fast after the first run (in the first run, code generation is the bottleneck, which is single threaded for now).

**Comment by azev77** [comment]
@s-broda I'm embarrassed to say I missed that part of the docs. I see it now & it's great! 

A table like this in the docs might help other users better see the "menu" of options:
![image](https://user-images.githubusercontent.com/7883904/85238687-b3088800-b3fd-11ea-8b80-d46e07cea506.png)
It might even help to clarify that: 
all conditional means are nested in `ARMA`
all conditional vol models (except `EGARCH`) are nested in `TGARCH`

Turns out you have a lot more than `Auto ARCH` you have `Auto E/TGARCH{o,p,q}-ARMA{p,q}`.
`auto.arma` is a special case: 
`auto.arma(df, bic)=selectmodel(ARCH{0}, df, meanspec=ARMA, criterion=bic)`

I see the [source here](https://github.com/s-broda/ARCHModels.jl/blob/e75597cecba11ad5357425e229ecb51a71c37f41/src/univariatearchmodel.jl#L419-L457) but I'm not sure I fully understand yet.

Below I will call model tuning parameters HP (hyper-parameters). 
Some questions-Feature requests (not sure which ones are important/urgent): 
1. currently the most general conditional mean model is `ARMA{p,q}`. 
Would it be possible to incorporate more general mean models? 
[TSAnalysis](https://github.com/fipelle/TSAnalysis.jl).jl has `varima`
There is also ARFIMA.jl https://github.com/fipelle/TSAnalysis.jl/issues/8
1. currently, lags for **ALL** HP are `1:maxlags`.
is it possible to allow `0:maxlags`? (then `intercept` is a special case of `ARMA`)
is it possible to have different ranges for different HP: p `0:2`,  q `1:2`?
currently it searches over ALL combinations of lags, `auto.arima` has `stepwise=true` allowing stepwise selection, which is faster
2. what is the current resampling strategy? 
Is it out-of-sample K-fold cross-validation? (I don't understand the code here) 
It could be nice to allow the user to select a custom resampling strategy as well as a custom criterion function...
3. suppose the user wants to auto select the "best" from all models in ARCHModels.jl, this is currently impossible bc EGARCH is not nested in TGARCH. So the user would need to store the IC from each model selection exercise.
Is it possible to allow the user to automatically include all models in `selectmodel`?

**Comment by s-broda** [comment]
Hi @azev77 

I like the table. Would you be able to open a pull request? I think it would best fit [here](https://s-broda.github.io/ARCHModels.jl/dev/univariatetypehierarchy/#Univariate-1). The code for this is [here](https://github.com/s-broda/ARCHModels.jl/blob/master/docs/src/univariatetypehierarchy.md). Otherwise, you could post the markdown code and I'll add it (it seems to be included as a picture in your code).

About your other points:

1. That's certainly possible, but unlikely to happen anytime soon unless someone else creates a PR.
2. Good point. The reason I excluded zero is that the resulting model is not always identified (take for example a `GARCH{p, q}` with `q=0`), which can lead to estimation problems. But there are valid reasons to want to do that. I've just opened a PR (#61) that introduces a `minlags` keyword argument. It defaults to 1, but you can set it to zero, or to something larger than 1 to get only a subset. Non-"rectangular" regions are unlikely to happen, because it wouldn't dovetail with the way I've implemented `selectmodel` based on `CartesianIndices`.
3. The package doesn't do resampling. Not sure what gives you the idea?
4. I'd rather not have a built-in method for this; I think it's better to give control to the user, because of the amount of models that can result. Note that you can do
```
julia> models = [selectmodel(VS, BG96; dist=D, minlags=0, maxlags=1) for VS in subtypes(UnivariateVolatilitySpec), D in setdiff(subtypes(StandardizedDistribution), [Standardized])];

julia> best_model = models[findmin(bic.(models))[2]]
```
This will minimize the BIC over all lag lengths, distributions, and GARCH model classes. I've added this to the documentation in #61.

5. I've implemented multi-step forecasts in #61 (except for the VaR, which requires Monte Carlo in the general case). Regarding returning the conditional distribution, I'll have to think about how the API for this would look, because `StandardizedDistribution` doesn't carry the mean and variance obviously. To get the conditional distribution, we'd need a type that wraps a `StandardizedDistribution` together with the mean and variance.  But then to do anything with it, we'd need to implement a host of methods (pdf, cdf, etc). Alternatively, this could be special cased for all included distributions and just return the corresponding unstandardized distribution from Distributions.jl (I don't think they have the GED though?). Note also that like for the VaR, this would only work one-step ahead without Monte Carlo. 

6. Thanks. Adding additional models is very simple, because the package is fully modular. I just need to find the time for it. Or you could give it a shot :) Just create a type that holds the coefficients in a vector, and implement the following methods for it: `nparams`, `presample`, `update`, `uncond`, `startingvals`, `constraints`, and `coefnames`. The meat of it is in `update!`, which implements the GARCH recursion. This will give you everything else (estimation, simulation, standard errors, display etc.) for free. You can use [this file](https://github.com/s-broda/ARCHModels.jl/blob/master/src/EGARCH.jl), which does this for EGARCH, for guidance.

Once CI passes, I'll merge #61 and release version 1.1. Would be cool if you could try out the new features (multi-step forecasting and minlags).

Thanks for your input!
Simon

**Comment by azev77** [comment]
@s-broda thank you for your response. 
I began working on a PR https://github.com/s-broda/ARCHModels.jl/pull/62
I'm not 
Is it possible for me to create a table in latex and paste my latex code in to the md file?

I think the following works:

| ``\mu_t`` 	| ``\sigma_t`` 	| `F`: Error Distribution 	|
|-	|-	|-	|
| `NoIntercept` 	| `ARCH{0}`: Homoscedastic 	| `StdNormal` 	|
| `Intercept` 	| `ARCH{q}` 	| `StdT` 	|
| `ARMA{p,q}` 	| `GARCH{p,q}` 	| `StdGED` 	|
| `Regression(x)` 	| `TGARCH{o,p,q}` 	| Std User-Defined 	|
|  	| `EGARCH{o,p,q}` 	|  	|

**Comment by azev77** [comment]
Let me clarify what I meant by resampling above (my point # 3). 

The updated list of potential feature requests: 
1. `Volatility specifications`: 
eg these models from [Hansen & Lunde 2005](https://onlinelibrary.wiley.com/doi/full/10.1002/jae.800) are currently not implemented
- [ ] IGARCH ![image](https://user-images.githubusercontent.com/7883904/85616628-06601d80-b62c-11ea-8c2e-1256e58b3663.png)
- [ ] Taylor/Schwert ![image](https://user-images.githubusercontent.com/7883904/85616827-46bf9b80-b62c-11ea-8a2c-227d606eae87.png)
- [ ] A-GARCH ![image](https://user-images.githubusercontent.com/7883904/85617094-af0e7d00-b62c-11ea-8a9e-87970e33533c.png)
- [ ] NA-GARCH 
- [ ] V-GARCH 
- [ ] GJR-GARCH
- [ ] log-GARCH
- [ ] N-GARCH (special case of A-PARCH w/o leverage effect)
- [ ] A-PARCH
- [ ] GQ-ARCH
- [ ] H-GARCH
- [ ] Aug-Garch


2. `Mean specifications`: `varima` (e.g. [TSAnalysis](https://github.com/fipelle/TSAnalysis.jl).jl), `varfima` (e.g. [ARFIMA](https://github.com/JuliaDynamics/ARFIMA.jl).jl), `varfimax`

**Comment by s-broda** [comment]
I agree that we'd need to use cross validation to get an unbiased estimate of the BIC, but as far as I can tell that's usually not what people do in this literature; they just pick the model that minimizes the BIC (or AIC) in-sample. Also, these data aren't IID, so you'd need to do something fancier than just K-fold CV. In principle this could be implemented as a scoring function that could be passed to `selectmodel`.

Thanks for the wishlist! I'll see what I can add in the future. My funding for working on this has run out, so I don't have as much time for working on this. PRs are welcome though!

**Comment by s-broda** [comment]
Note that GJR is the same as TGARCH, so this is already implemented.

**Comment by azev77** [comment]
I see, bc in the paper, they're defined separately:
![image](https://user-images.githubusercontent.com/7883904/86027085-40e80280-b9fe-11ea-973a-262cf695f278.png)

**Comment by s-broda** [comment]
Those two are clearly not the same (different power of sigma). Usually TGARCH is understood to be what's called GJR above.

**Comment by azev77** [comment]
Here is their table 1 of specifications for the conditional variance: 
![image](https://user-images.githubusercontent.com/7883904/86027424-afc55b80-b9fe-11ea-94e3-1e232d3948d7.png)

**Comment by azev77** [comment]
Btw, it's common for packages to show off benchmarks (w/ R & Python) in the Readme: 
"We find that ARCHModels.jl is faster than Python’s ARCH package by a factor of about 6, faster than Matlab’s Econometrics Toolbox by a factor of about 12, and faster than R’s rugarch package by a factor of about 50. This is despite the fact that all packages (except ARCHModels.jl) implement the likelihood function in a compiled low-level language."

Or even better, add a nice benchmark figure like [here](https://github.com/PyDataBlog/ParallelKMeans.jl).

**Comment by s-broda** [comment]
Yeah, I had thought about that. The best would be to have the benchmarks run on CI automatically, but that's not a good idea because you're not guaranteed a specific machine there.

**Comment by azev77** [comment]
@s-broda 
I'd like to work on adding `family GARCH` from the list above, but not sure where to start. 
Can you help me get started with this?

**Comment by s-broda** [comment]
Hi @azev77 

Sure! Basically you just need to follow the structure in TGARCH.jl.  I.e. (using `o` instead of `m` for consistency with the rest of the code), you start buy defining a  `struct FamGARCH{o, p, q, T<:AbstractFloat} <: UnivariateVolatilitySpec{T}`. This struct should contain a vector for the model parameters, which should appear in the order in which they relate to the type parameters `o`, `p`, `q`.

You then implement the following functions:

 * `nparams`: number of parameters given `o`, `p`, and `q` ( I guess ``1+ o + p + 3q`` in this case?)
 * `presample`: how many presample values are needed for conditional MLE (usually `max(o, p, q)`)
 * `update!`: the actual GARCH recursion
 * `uncond`: the unconditional variance given the parameters
 * `startingvals`: the starting values for all parameters
 * `constraints`: necessary (but usually not sufficient) constraints that keep the model in the stationary region.
 * `subsetmask` is related to subset GARCH models: it takes a large model (say `GARCH{2, 2}`)  and a tuple (say `(1, 1)`) representing a smaller model, and returns a `BitArray` that has `false` wherever the parameter in the large model is zero for the subset model, e.g., 
```julia
julia> mask =  ARCHModels.subsetmask(GARCH{2, 2}, (1, 1))
5-element BitArray{1}:
 1
 1
 0
 1
 0
```
 * `subsettuple` is basically the inverse function to `subsetmask`. It takes a `BitArray` and returns the tuple:
```julia
julia> ARCHModels.subsettuple(TGARCH{2, 3, 3}, mask)
(1, 1, 0)
```

Have fun!
Simon

**Comment by azev77** [comment]
@s-broda you may have seen this, the textbook [Financial Risk Forecasting ](https://www.financialriskforecasting.com/book-code/) now uses ARCHModels.jl. 
They have blanks for: 
[APARCH](https://www.financialriskforecasting.com/code/PythonJulia2.html)
[OGARCH ](https://www.financialriskforecasting.com/code/PythonJulia3.html)

**Comment by s-broda** [comment]
Cool, thanks for the pointer!

APARCH should be coming soon.

**Comment by miromarszal** [comment]
Hello, first thanks for all the hard work that went into this great package! I was wondering if allowing to fit ARMA models directly would still fit within the scope. I know that I can do:
```julia
fit(ARCH{0}, data; meanspec=ARMA{p,q})
```
but a wrapper like this would be very neat:
```julia
fit(ARMA{p,q}, data)
```
I'm saying this since it took me a minute to figure this out from the docs and I was close to moving on ;)
Also, it would be nice to allow specifying no intercept in ARMA, maybe like this:
```julia
fit(ARMA{p,q}, data; meanspec=NoIntercept())
```

**Comment by s-broda** [comment]
Thanks for the suggestion. The convenience methods for `fit(::Type{<:ARMA}, data` (and `selectmodel`) have been implemented (https://github.com/s-broda/ARCHModels.jl/commit/99b389f677edb1f43f2bfc31176ff2c635ea36dc) and and will make it into 2.2. 

Fitting ARMA models w/o intercept the way you suggest would require a bigger rewrite, so this is unlikely to happen. It would be quite easy though to implement an additional `meanspec` (`ZeroMeanARMA`?) that does this. Not sure if that would be worth it though, in my experience it is pretty rare that one needs this.
